File: app/src/main/java/com/liadpaz/music/MainActivity.java
package com.liadpaz.music;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.provider.DocumentsContract;
import android.view.Menu;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.viewpager2.widget.ViewPager2;

import com.google.android.material.tabs.TabLayoutMediator;
import com.liadpaz.music.adapters.ViewPagerAdapter;
import com.liadpaz.music.databinding.ActivityMainBinding;
import com.liadpaz.music.fragments.BlankFragment;
import com.liadpaz.music.fragments.SongsListFragment;
import com.liadpaz.music.utils.LocalFiles;
import com.liadpaz.music.utils.Utilities;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {

    private static final int REQUEST_PERMISSION = 459;
    private static final int REQUEST_PICK_FOLDER = 44;
    @SuppressWarnings("unused")
    private static final String TAG = "MAIN_ACTIVITY";

    private ActivityMainBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        setSupportActionBar(binding.toolBarMain);

        new LocalFiles(getSharedPreferences("Music.Data", 0));

        if (PackageManager.PERMISSION_GRANTED != ContextCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE)) {
            requestPermissions(new String[]{Manifest.permission.READ_EXTERNAL_STORAGE}, REQUEST_PERMISSION);
        } else {
            setViewPager();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menuItemSelectFolder: {
                startActivityForResult(new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE), REQUEST_PICK_FOLDER);
                break;
            }

            case R.id.menuItemSettings: {
                //TODO: add settings activity
                break;
            }

            case R.id.menuItemAbout: {
                //TODO: add about activity
                break;
            }

            default: {
                return false;
            }
        }
        return true;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (requestCode == REQUEST_PERMISSION) {
            setViewPager();
        }
    }

    private void setViewPager() {
        ViewPager2 viewPager = binding.viewPagerMain;

        viewPager.setAdapter(new ViewPagerAdapter(this, new ArrayList<Class>() {{
            add(SongsListFragment.class);
            add(BlankFragment.class);
        }}));

        ArrayList<String> tabsTitle = new ArrayList<String>() {{
            add(getString(R.string.tab_songs));
            add("Blank");
        }};

        new TabLayoutMediator(binding.tabLayoutMain, viewPager, ((tab, position) -> tab.setText(tabsTitle.get(position)))).attach();
    }

    @SuppressWarnings("ConstantConditions")
    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == RESULT_OK && requestCode == REQUEST_PICK_FOLDER) {
            String path = Utilities.getPathFromUri(DocumentsContract.buildDocumentUriUsingTree(data.getData(), DocumentsContract.getTreeDocumentId(data.getData())));
            if (!LocalFiles.getPath().equals(path)) {
                LocalFiles.setPath(path);
                recreate();
            }
        }
    }

    @Override
    public void recreate() {
        overridePendingTransition(0,0);
        finish();
        overridePendingTransition(0,0);
        startActivity(getIntent());
    }
}

File: app/src/main/java/com/liadpaz/amp/ui/MainActivity.kt
package com.liadpaz.amp.ui

import android.app.Activity
import android.app.SearchManager
import android.content.Context
import android.content.Intent
import android.media.AudioManager
import android.os.Bundle
import android.util.TypedValue
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.SearchView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.liadpaz.amp.R
import com.liadpaz.amp.databinding.ActivityMainBinding
import com.liadpaz.amp.livedatautils.SongsUtil.getSongs
import com.liadpaz.amp.ui.fragments.ExtendedFragment
import com.liadpaz.amp.ui.fragments.MainViewPagerFragment
import com.liadpaz.amp.ui.fragments.SearchFragment.Companion.newInstance
import com.liadpaz.amp.ui.viewmodels.Song
import com.liadpaz.amp.utils.Constants
import com.liadpaz.amp.utils.LocalFiles.showCurrent
import java.util.*
import kotlin.collections.ArrayList

class MainActivity : AppCompatActivity() {
    
    lateinit var binding: ActivityMainBinding

    private var searchItem: MenuItem? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
        setContentView(ActivityMainBinding.inflate(layoutInflater).also { binding = it }.root)
        setSupportActionBar(binding.toolBarMain)

        volumeControlStream = AudioManager.STREAM_MUSIC

        supportFragmentManager.beginTransaction().replace(R.id.mainFragment, MainViewPagerFragment.newInstance()).replace(R.id.extendedFragment, ExtendedFragment.newInstance()).commitNow()
        if (intent != null) {
            if (intent.hasExtra(Constants.PREFERENCES_SHOW_CURRENT) && showCurrent) {
                BottomSheetBehavior.from(binding.extendedFragment).state = BottomSheetBehavior.STATE_EXPANDED
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        (menu.findItem(R.id.menuSearch).also { searchItem = it }.actionView as SearchView).setSearchableInfo((getSystemService(Context.SEARCH_SERVICE) as SearchManager).getSearchableInfo(componentName))
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean =
            when (item.itemId) {
                R.id.menuItemSettings -> {
                    startActivityForResult(Intent(applicationContext, SettingsActivity::class.java), REQUEST_SETTINGS)
                    true
                }
                R.id.menuItemAbout -> {
                    startActivity(Intent(applicationContext, AboutActivity::class.java))
                    true
                }
                else -> false
            }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) =
            if (requestCode == REQUEST_SETTINGS && resultCode == Activity.RESULT_OK) {
                recreate()
            } else {
                super.onActivityResult(requestCode, resultCode, data)
            }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        if (intent.hasExtra(Constants.PREFERENCES_SHOW_CURRENT) && showCurrent) {
            BottomSheetBehavior.from(binding.extendedFragment).setState(BottomSheetBehavior.STATE_EXPANDED)
        } else {
            handleIntent(intent)
        }
    }

    override fun recreate() {
        overridePendingTransition(0, 0)
        finish()
        overridePendingTransition(0, 0)
        startActivity(intent)
    }

    /**
     * This function handles the query intent.
     *
     * @param intent An [Intent] with `Intent.ACTION_SEARCH` as its action and a `SearchManager.QUERY` extra as the query parameter.
     */
    private fun handleIntent(intent: Intent?) {
        if (intent != null && Intent.ACTION_SEARCH == intent.action && intent.hasExtra(SearchManager.QUERY)) {
            query(intent.getStringExtra(SearchManager.QUERY)!!)
        }
    }

    /**
     * This function queries the list of songs corresponds to the query string.
     *
     * @param queryString The query string.
     */
    private fun query(queryString: String) {
        val finalQuery = queryString.toLowerCase(Locale.US)
        val queriedSongs: ArrayList<Song> = ArrayList(getSongs().filter { song -> song.isMatchingQuery(finalQuery) })
        supportFragmentManager.beginTransaction().replace(R.id.mainFragment, newInstance(queryString, queriedSongs)).addToBackStack(null).commitNow()
    }

    fun setBottomSheetHidden(isHidden: Boolean) {
        val layoutParams = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        val bottomSheetBehavior = BottomSheetBehavior.from<View>(binding.extendedFragment)
        if (isHidden) {
            layoutParams.setMargins(0, 0, 0, 0)
            bottomSheetBehavior.isHideable = true
            bottomSheetBehavior.setState(BottomSheetBehavior.STATE_HIDDEN)
        } else {
            layoutParams.setMargins(0, 0, 0, TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 72f, resources.displayMetrics).toInt())
            bottomSheetBehavior.state = if (bottomSheetBehavior.state == BottomSheetBehavior.STATE_HIDDEN) BottomSheetBehavior.STATE_COLLAPSED else bottomSheetBehavior.state
            bottomSheetBehavior.setHideable(false)
        }
        binding.mainFragment.layoutParams = layoutParams
    }

    override fun onBackPressed(): Unit = with(BottomSheetBehavior.from<View>(binding.extendedFragment)) {
        when {
            state == BottomSheetBehavior.STATE_EXPANDED -> {
                state = BottomSheetBehavior.STATE_COLLAPSED
            }
            searchItem!!.isActionViewExpanded -> {
                searchItem!!.collapseActionView()
            }
            else -> {
                super.onBackPressed()
            }
        }
    }

    companion object {
        private const val TAG = "AmpApp.MainActivity"
        private const val REQUEST_SETTINGS = 525
    }
}